### ISSUE TITLE
Issue when manipulating kernel_product's output

### ISSUE THREAD

**Issue by SamuelJoutard** [getkeops/keops#77]
Hi,

I am having issues when I want to do some operations on kernel_product's outputs. Here is a code to reproduce the error I get.

```python
import torch
from pykeops.torch import Kernel, kernel_product
from pykeops.torch import LazyTensor

sigma_f = torch.tensor([1.]).type(torch.FloatTensor).cuda()
sigma_l = torch.tensor([1.]).type(torch.FloatTensor).cuda()

params = {
        "id"      : Kernel("gaussian(x,y) * gaussian(u,v)"),
        "gamma"   : (.5/sigma_f**2, .5/sigma_l**2),
}

N = 10000
M = 20000

y = torch.rand((N, 5)).cuda().contiguous()
y.requires_grad = True

x = torch.rand((M, 5)).cuda().contiguous()
x.requires_grad = True

K_points = torch.rand((N, 4)).cuda().contiguous()
grid_x = torch.rand((M, 4)).cuda().contiguous()

K_points = kernel_product(params, (y, K_points), (x, grid_x), grid_x)
q = (K_points[:, 3:].clone() + 0.0000000001)
K_points[:, :3] = K_points[:, :3].clone() / q

loss = torch.dot(K_points.view(-1), torch.ones_like(K_points).view(-1))
loss.backward()
```

This throw the following error:

> Traceback (most recent call last):
  File "Code/non_local_feature_alignment/test_keops.py", line 55, in <module>
    loss.backward()
  File "/opt/conda/lib/python3.6/site-packages/torch/tensor.py", line 198, in backward
    torch.autograd.backward(self, gradient, retain_graph, create_graph)
  File "/opt/conda/lib/python3.6/site-packages/torch/autograd/__init__.py", line 99, in backward
    allow_unreachable=True)  # allow_unreachable flag
RuntimeError: one of the variables needed for gradient computation has been modified by an inplace operation: [torch.cuda.FloatTensor [10000, 4]], which is output 0 of GenredAutogradBackward, is at version 1; expected version 0 instead. Hint: enable anomaly detection to find the operation that failed to compute its gradient, with torch.autograd.set_detect_anomaly(True). 

This error does not appear if you comment the lines:
```python
q = (K_points[:, 3:].clone() + 0.0000000001)
K_points[:, :3] = K_points[:, :3].clone() / q
```

Is the output of kernel_product supposed to be handled in a specific way? I would like here to normalize the output of the kernel product by the sum of each row of the kernel so in my code the last dimension of `grid_x` only contains ones (same for K_points).

**Comment by SamuelJoutard** [comment]
Hi again,

I found a solution to do what I wanted to do without raising this error by changing the mode argument.

```
K_points = torch.rand((N, 3)).cuda().contiguous()
K_points_lazy = LazyTensor(K_points[:, None, :])
grid_x = torch.rand((M, 3)).cuda().contiguous()
grid_x_lazy = LazyTensor(grid_x[None, :, :])

q = (-((y_lazy - x_lazy)**2).sum(dim=2)/2 - ((grid_x_lazy - K_points_lazy)**2).sum(dim=2)/2).logsumexp(dim=1)

res = kernel_product(params, 
                    (y, K_points), 
                    (x, grid_x), 
                    grid_x, 
                    -q,
                    torch.zeros((M, 1)).cuda(),
                    mode="log_scaled" 
                    )

loss = torch.dot(res.view(-1), torch.ones_like(res).view(-1))

print(loss)
loss.backward()
```

It might still be worth to have a look at the previous code just in case the previous error was related to PyKeops.

Thanks anyway,

Samuel

**Comment by jeanfeydy** [comment]
Hi @SamuelJoutard ,

Happy to see that you solved the problem quickly :-) 
I have two remarks:

1. As far as I can tell, your initial problem was not related to KeOps but to the `torch.autograd` engine itself. As explained for instance in [our documentation](http://www.kernel-operations.io/keops/autodiff_gpus/backpropagation.html#id1), the backpropogation algorithm must keep in memory all the intermediate results leading to a computation. Since in-place affectations modify buffers without re-allocating new ones (in your case, the `K_points[:, :3] = ... ` statement), they are poorly supported by automatic differentiation engines: this is discussed explicitly in the [PyTorch documentation](https://pytorch.org/docs/stable/notes/autograd.html#in-place-operations-with-autograd). A simple workaround is to compute the updated version of `K_points` using "constructive" operations, such as `torch.cat`.

2. The `kernel_product` helper has been made somewhat obsolete by the `LazyTensor` wrapper, and we were thinking about deprecating it in future versions. Would you be fine with this, or are there features of the `kernel_product` syntax that we should port to `LazyTensors` first, in your opinion? Feedback is always welcome!

Best regards - and see you soon,
Jean

**Comment by SamuelJoutard** [comment]
Hi Jean,

Thank you for your detailed answer. 

Regarding your second point, I actually switched to `sumsoftmaxweight` reduction on Lazytensor as I assume it better handles numerical issues related to quotients of small exponentials. So right now I can't think of a particular use case of kernel_product being more appropriate than Lazytensors in my work. 

That being said, it raised another issue related to gradient arguments not being contiguous (I already replaced all sums by a dot product):

> RuntimeError: [Keops] Arg at position 4: is not contiguous. Please provide 'contiguous' dara array, as KeOps does not support strides. If you're getting this error in the 'backward' pass of a code using torch.sum() on the output of a KeOps routine, you should consider replacing 'a.sum()' with 'torch.dot(a.view(-1), torch.ones_like(a).view(-1))'.

But I will raise another issue with some code to reproduce if I can't solve this. 

Thanks again for the help,

Best regards,

Samuel
